### PR TITLE
[CSL-2014] Report misbehaviors to EKG metrics

### PR DESCRIPTION
**Pull request by hithroc**

### CODE REVIEW

File: infra/Pos/Reporting/MemState.hs [hunk 1]
```diff
@@ -2,21 +2,39 @@
...
+import           System.Metrics.Counter (Counter)
 import           Control.Lens (makeLenses)
 import           System.Wlog.LoggerConfig (LoggerConfig)
 import           Universum
 
+-- | EKG metric values for misbehaviors
+data MisbehaviorMetrics = MisbehaviorMetrics
+    { _mmRollbacks :: Gauge
+    , _mmSscFailures :: Counter
+    , _mmIgnoredCommitments :: Gauge
```
Comment (gromakovsky) [inline infra/Pos/Reporting/MemState.hs]: Can you document all three fields please?

File: block/src/Pos/Block/Network/Logic.hs [hunk 2]
```diff
@@ -561,6 +563,11 @@ applyWithRollback nodeId enqueue toApply lca toRollback = do
     reportRollback = do
         let rollbackDepth = length toRollback
         let isCritical = rollbackDepth >= criticalForkThreshold
+
+        -- Commit rollback value to EKG
+        whenJustM (view misbehaviorMetrics) $ liftIO .
+            flip Metrics.set (fromIntegral rollbackDepth) . _mmRollbacks
```
Comment (gromakovsky) [inline block/src/Pos/Block/Network/Logic.hs]: But wait, the issue description suggests that `_mmRollbacks` gauge corresponds to
> the maximal rollback depth in last T (where is configurable time value, e. g. few hours).

File: infra/Pos/Reporting/Methods.hs [hunk 3]
```diff
@@ -343,7 +343,8 @@ reportNode sendLogs extendWithNodeInfo reportType =
 reportMisbehaviour
     :: forall ctx m . (MonadReporting ctx m)
     => Bool -> Text -> m ()
-reportMisbehaviour isCritical message = do
+reportMisbehaviour isCritical message = when False $ do -- TODO: Add a flag wether to report
```
Comment (gromakovsky) [inline infra/Pos/Reporting/Methods.hs]: You could write that famous
```
reportMisbehaviour isCritical message = don't $ do
    bla
  where
    don't _ = pure ()
```

:)

File: lib/src/Pos/Launcher/Resource.hs [hunk 4]
```diff
@@ -293,6 +295,12 @@ allocateNodeContext ancd txpSettings = do
     let slotDuration :: Integer
         slotDuration = toMicroseconds . bvdSlotDuration $ gdBlockVersionData genesisData
     ncSubscriptionKeepAliveTimer <- newTimer $ 3 * fromIntegral slotDuration
+
+    _mmRollbacks <- liftIO $ Metrics.createGauge "cardano.BlockRollbacks" ekgStore
+    _mmSscFailures <- liftIO $ Metrics.createCounter "cardano.SCCComputationFailures" ekgStore
+    _mmIgnoredCommitments <- liftIO $ Metrics.createGauge "cardano.IgnoredCommitments" ekgStore
+    let mm = MisbehaviorMetrics {..}
```
Comment (gromakovsky) [inline lib/src/Pos/Launcher/Resource.hs]: I'd define `MonadIO m => EKG.Store -> m MisbehaviorMetrics` in `Pos.Reporting.MemState` (near `MisbehaviorMetrics` definition).
Also please use `Pos.System.Metrics.Constants.withCardanoNamespace` for metric names.

File: lib/src/Pos/Launcher/Resource.hs [hunk 5]
```diff
@@ -301,6 +309,7 @@ allocateNodeContext ancd txpSettings = do
             , ncNodeParams = np
             , ncTxpGlobalSettings = txpSettings
             , ncNetworkConfig = networkConfig
+            , ncMisbehaviorMetrics = Just mm
```
Comment (gromakovsky) [inline lib/src/Pos/Launcher/Resource.hs]: I might be missing something, but it looks like `ncMisbehaviorMetrics` is always `Just`. Why is it `Maybe`?

Comment (hithroc) [reply lib/src/Pos/Launcher/Resource.hs]: `Pos.Reporting.MemState` has `emptyReportingContext :: ReportingContext`, which is used in some tests and other things that I haven't dived into. That's where `Nothing` for misbehavior metrics is used.

File: block/src/Pos/Block/Network/Logic.hs (hunk 2 above)
Comment (gromakovsky) [reply block/src/Pos/Block/Network/Logic.hs]: We decided that we don't need it and this logic is ok.

File: lib/src/Pos/Launcher/Resource.hs (hunk 5 above)
Comment (gromakovsky) [reply lib/src/Pos/Launcher/Resource.hs]: Ah, so it's needed for `setter`, ok!

**Pull request merged**